API/hermes_abi/hermes_abi.h

```diff
@@ -0,0 +1,400 @@
+  V(Object)                         \
+  V(Array)                          \
+  V(String)                         \
+  V(BigInt)                         \
+  V(Symbol)                         \
+  V(Function)                       \
+  V(ArrayBuffer)                    \
+  V(PropNameID)                     \
+  V(WeakObject)
+
+#define DECLARE_HERMES_ABI_POINTER_TYPE(name) \
+  struct HermesABI##name {                    \
+    HermesABIManagedPointer *pointer;         \
+  };                                          \
+  struct HermesABI##name##OrError {           \
+    uintptr_t ptrOrError;                     \
+  };
+
+HERMES_ABI_POINTER_TYPES(DECLARE_HERMES_ABI_POINTER_TYPE)
+#undef DECLARE_HERMES_ABI_POINTER_TYPE
+
+#define HERMES_ABI_TRIVIAL_OR_ERROR_TYPES(V) \
+  V(Bool, bool)                              \
+  V(Uint64, uint64_t)                        \
+  V(Uint8Ptr, uint8_t *)                     \
+  V(SizeT, size_t)                           \
+  V(PropNameIDListPtr, HermesABIPropNameIDList *)
+
+#define DECLARE_HERMES_ABI_TRIVIAL_OR_ERROR_TYPES(name, type) \
+  struct HermesABI##name##OrError {                           \
+    bool is_error;                                            \
+    union {                                                   \
+      type val;                                               \
+      uint16_t error;                                         \
+    } data;                                                   \
+  };
+
+HERMES_ABI_TRIVIAL_OR_ERROR_TYPES(DECLARE_HERMES_ABI_TRIVIAL_OR_ERROR_TYPES)
+#undef DECLARE_HERMES_ABI_TRIVIAL_OR_ERROR_TYPES
+
+struct HermesABIVoidOrError {
+  bool is_error;
+  uint16_t error;
+};
+
+/// Always set the top bit for pointers so they can be easily checked.
+#define HERMES_ABI_POINTER_MASK (1 << (sizeof(int) * 8 - 1))
+
+enum HermesABIValueKind {
+  HermesABIValueKindUndefined = 0,
+  HermesABIValueKindNull = 1,
+  HermesABIValueKindBoolean = 2,
+  HermesABIValueKindError = 3,
+  HermesABIValueKindNumber = 4,
+  HermesABIValueKindSymbol = 5 | HERMES_ABI_POINTER_MASK,
+  HermesABIValueKindBigInt = 6 | HERMES_ABI_POINTER_MASK,
+  HermesABIValueKindString = 7 | HERMES_ABI_POINTER_MASK,
+  HermesABIValueKindObject = 9 | HERMES_ABI_POINTER_MASK,
+};
+
+struct HermesABIValue {
+  HermesABIValueKind kind;
+  union {
+    bool boolean;
+    double number;
+    HermesABIManagedPointer *pointer;
+    HermesABIErrorCode error;
+  } data;
+};
+struct HermesABIValueOrError {
+  HermesABIValue value;
+};
+
+struct HermesABIASCIIRef {
+  const char *data;
+  size_t length;
+};
+
+/// Define the structure for buffers containing JS source or bytecode. This is
+/// designed to mirror the functionality of jsi::Buffer.
+struct HermesABIBufferVTable {
+  void (*release)(HermesABIBuffer *);
+};
+struct HermesABIBuffer {
+  const HermesABIBufferVTable *vtable;
+  const uint8_t *data;
+  size_t size;
+};
+
+struct HermesABIByteBufferVTable {
+  void (*grow_by)(HermesABIByteBuffer *, size_t);
```

Comment:
What if the buffer cannot grow and we get out of memory error?
We cannot propagate an exception here, but should we return bool or an error code?

Reply:
I considered it, and the API does permit allocation failure  (you just wouldn't update `available` if the allocation fails). 

At the same time, correctly detecting and handling OOMs is extremely difficult. Even more so when we're using non-exception compiled code as is the case in the ABI implementation. Given that it is going to be rarely useful in practice, I don't think native OOM handling should be a first-class design consideration.

What do you think?

Reply:
I think you have a good point about not updating the `available` field. No need to provide any extra result then.
Please make sure that this case is handled correctly on both ends of the API.

Reply:
@neildhar  can you provide an example usage of `HermesABIByteBuffer`? is this managed by hermes or you expect the caller to manage this object?

Reply:
For most Hermes integrators, this entire API will be an implementation detail. The intention is for this to provide a stable ABI boundary, and then for the integrator to use the C++ wrapper in `HermesABIRuntime.cpp` to use it as a normal JSI runtime.

Reply:
maybe i should have been more clear in my question. I did not find a factory function that creates and returns a  `HermesABIByteBuffer*` .